---
title: "Quick start guide"
output:
  pdf_document: default
  html_document:
    df_print: paged
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

# Quick start guide

## Run

1.  Load the library in R.

    `library(tugMedi)`

2.  Copy the test data of this package under the current directory.

    `copy_test_data("./tugMedi.1.0/")`

    You will get Test1/, Test2, ... under the current directory.

3.  Move into Test1/ and run a test.

    `setwd("Test1/")`

    `source("test1.R")`

    You will get a renewed Output/ under this test directory.

-   Alternatively, you can `q()` after `copy_test_data()`.

    Then, after moving into the test directory, on your OS,

    `Rscript --vanilla --slave test1.R`

\* Note that this is just an artificial example for explanatory purposes. **The inputs and outputs do not have any biological meanings**.

## test1.R

Please see test1.R, and edit it when you understand.

### Simulation

```         
start_simulation( input_files = 'Input/DATA/FILES.txt', 
                  saveTo      = 'Output/Results.sim.01.RDS', 
                  seed        = NA
                 )
```

This command starts a simulation. Input data are defined through option `input_files`. Please see the section of Input data for the details of input data. RDS data are saved to `saveTo` for the second simulation, where usually drug intervention is performed. Drug intervention is explained below. You can change the random seed through `seed`.

### Generators

These commands randomly generate parameter values in specific manners for Approximate Bayesian computation later. You can comment out these commands and manually set parameter values, too.

```         
write_prm(       input = './Input/ForGenerators/parameters.parameters.txt')
write_CF(        input = './Input/ForGenerators/CF.parameters.txt')
write_weights(   input = './Input/ForGenerators/hallmark_weights.parameters.txt' )
write_cloneinit( input  = './Input/ForGenerators/cloneinit.parameters.txt' )
write_Rint(      input = './Input/ForGenerators/Rint.parameters.txt' )
write_EF.Rint(   input = './Input/ForGenerators/EF.Rint.parameters.txt' )
# Only this depends on write_prm() result
write_EF.Rother( input = './Input/ForGenerators/EF.Rother.parameters.txt' )
```

The inputs of all these functions are independent except that of `write_EF.Rother()`, which depends on the output of `write_prm()`. So, please put `write_EF.Rother()` after `write_prm()`.

See the section of Generators for the details of these functions.

### Post processes

These commands take in the outputs of the simulation and perform post processes such as the calculation of VAF (variant allele frequency).

```         
write_VAF(   input = './Input/ForPosts/VAF.parameters.txt' )
write_TMB(   input = './Input/ForPosts/TMB.parameters.txt' )
write_evals( input = './Input/ForPosts/evals.parameters.txt' )
write_realTime_clone( input = './Input/ForPosts/realTime.parameters.txt' )
```

`write_TMB()` and `write_evals()` use the output of `write_VAF()`.

See the section of Post processes for the details of these functions.

## Drug intervention in test1.R

```         
start_simulation( first       = F,
                  input_files = 'Input/DATA/FILES.drgInt.txt', 
                  loadFrom    = 'Output/Results.sim.01.RDS', 
                  saveTo      = 'Output/Results.sim.02.RDS', 
                  seed        = NA,
                  drug_int_param = drug_int_param
                 )
```

One can simulate how moleculary targeted drugs take effect on cancer cells with this command. Because drug intervention simulation usually takes advantage of the results of the first simulation, option `first` is `F`. The results of the first simulation are loaded with `loadFrom`, and the results of drug intervention simulation are saved to `saveTo`. As with the first simulation, input data are defined through `input_files` and the random seed can be set with `seed`.

To continue from the first simulation, it is necessary to update the first RDS file before the second simulation by the command:

```         
update_RDS_from_output( in.file  = 'Input/DATA/FILES.txt', 
                        rds.file = 'Output/Results.sim.01.RDS' )
```

By this, simulation parameters can be changed through the input data at the second simulation. Typically, you may extend `control.censor_cell_number` (the censor [program stop] of the maximum cell number) and `control.censor_time_step` (maximum simulation time step) in the input data for drug intervention. (In this example, the second simulation command reads `input_files = 'Input/DATA/FILES.drgInt.txt'` to use parameters.[drgInt]{.underline}.txt, an input file different from parameters.txt, used in the first simulation.) The second simulation starts from the next to the last time step of the first simulation. The last state of clones can be changed through editing an output file named cloneout.txt. See Chapter 11 for further details.

`drug_int_param` in the second simulation command is set as follows:

```         
drug_int_param <- list( 'kill_prob' = 0.5,
                        'block_prob'= 0.0,
                        'gene' = c('KRAS', 'APC') ) # Any of them is a target
```

Cells with any of the genes indicated by `'gene'` in the dysfunctional state are the targets of "kill" or "block" operations. The kill operation kills cells at the rate of `'kill_prob'` per simulation time step. The block operation blocks the dysfunctional state of the genes, i.e., revert the dysfunctional state ('1', indicating the dysfunction) to the normal state ('0', the normal state) of the genes in the program. This blocking is performed on cells randomly selected at the rate of `'block_prob'` per simulation time step.

## Outputs

### Output/

You will get a new Output/, which has:

-   cloneout.txt: Simulation result of cancer-cell evolution.
-   cloneout.realTime.txt: Estimated real time is added to cloneout.txt.
-   Results.sim.01.RDS: RDS data for the first simulation.
-   Results.sim.02.RDS: RDS data for the second (drug intervention) simulation.
-   Mutations/: Data on point mutations and CNAs are stored.
-   VAF/: VAF data are stored.
-   Info/: Other information is stored.

(The older Output/ is automatically renamed with a time stamp.)

See the section of Output data for the details of these outputs.

### Cancer-cell evolution

In short, you can first see cloneout.txt. Simulation time step represented in column Time increases with the rows. Column N_primary, meaning the number of cells in the primary tumor, and N_metastatic, the number of cells in the metastatic tumor, may increase with the time or become 0 at some time point, the latter of which means extinction. Column 'ID', i.e., clone ID, shows various clones over the time, which means intra-tumor heterogeneity (ITH).

Columns genesMalfunc and genesWmutsNoExp show genes in the malfunctional state and genes with mutations not causing the malfunction, respectively. The latter genes are mostly passenger mutations, but the situation may be complex if the genes are defined as recessive. Under the recessive, such mutations can be driver mutations because recessive genes need to have driver mutations on both parental chromosomes to cause the malfunction.

Columns PointMut_ID and CNA_ID show the IDs of point mutations and CNAs, respectively. They are driver or passenger mutations. The detailed information is stored under Mutations/.

Different clones have different mutations (point mutations and CNAs). But, if clones are metastatic clones as shown in column Type, clones with the same mutations may appear because primary clones that keep the same mutations seed metastatic clones at different time steps, as distinguishable in column Birth_time.

Different clones have different trial probabilities (columns d to im) and different hallmark variable values (Hd to Hb).

### Drug intervention simulation

In cloneout.txt, you may see a decrease of N_primary or N_metastatic at last time steps. This results from the drug intervention simulation. Only clones with specified target genes decrease as shown in column N_cells, which represents the number of cells in each clone.
