---
title: "Preparing observed data"
output:
  pdf_document: default
  html_document:
    df_print: paged
vignette: >
  %\VignetteEncoding{UTF-8}
  %\VignetteEngine{knitr::rmarkdown}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
  
)
```

# Preparing observed data

Observed data are not required by the simulation itselt, but they are necessary for ABC and evaluations and useful for setting simulation inputs.

## See observed SNV/indel data you have

For ABC and evaluations, it is necessary to format observed data into samples.{Rint, Rother}.txt, which only have to have the columns of gene name, sample name, and VAFs.

-   *You can change the names of samples.{Rint, Rother}.txt, but here we use these names for convenience.*

### 0) Confirm if observed SNVs/indels are tumor-specific.

-   To see cancer cells evolving from normal cells, SNVs/indels must be tumor-specific, that is, the differences of cancer cells from normal cells.

### 1) Add the VAF column

-   Calculate observed VAF if it's absent, and add the column.

    -   e.g., VAF = t_alt_count / ( t_alt_count + t_ref_count ) in the MAF format.

### 2) Select columns

-   Keep the columns of gene name and sample name.

-   The column numbers of the three columns need to be later specified.

    -   Hence, the order does not matter.

-   Keep other columns useful for interpretation. These columns are not necessary to run the program, but will be informative later.

    -   For the MAF format, we usually select the columns of Hugo_Symbol, Chromosome, Start_Position, End_Position, Variant_Type, Tumor_Sample_Barcode, HGVSc, HGVSp_Short, t_ref_count, and t_alt_count, adding t_vaf (calculated VAF).

### 3) Select samples

-   You can limit samples if the file size is large. A sample among them will be later used for analysis.

### 4-a) Set samples.Rint.txt – regions of interest

-   From the file of subsection 3), select rows with possible driver genes you think in the column of gene name.

    -   You can sort genes with VAFs. Generally, well-known tumor-related genes with high VAFs are likely to be drivers.

    -   It is likely that genes with low VAFs close to LOD have little influence, if any.

    -   You can later evaluate how much your hypothesis is likely through the model comparison technique of ABC.

        *\* If there are multiple SNVs/indels for a possible driver gene, you may select one with the highest VAF. The others can be threw away or kept as passengers.*

### 4-b) Set samples.Rother.txt – other regions

-   From the file of subsection 3), select rows with possible passenger genes you think in the column of gene name.

    -   You can sort chromosome numbers and then keep rows until you think the number of passengers is enough (e.g., around 10 at the first trial).

-   You many need to change the gene names to artificial gene names to match with artificial gene names in Input/Rother.txt.

## See observed CNA data if you have

Observed CNA data can be useful for setting simulation inputs.

*\* If you do not have, you may apply data of the same cancer type you have already had to the estimates below as priors.*

### 1) Estimation of m_del and m_dup

*\* m_del and m_dup are in Input/parameters.txt.*

-   Count SNVs/indels across autosomes.

-   Count dels and dups across autosomes.

    -   You may need to see a histogram of logR values (associated with copy numbers)

        and then decide a cut-off value to determine del and dup.

-   Take the ratio of the number of dels to that of SNVs/indels, which equals to the ratio of m_del to m0.

-   Apply the same to dups.

-   Then, you can get the estimates of m_del and m_dup to a specific m0.

### 2) Estimation of ave_len_del and ave_len_dup

*\* ave_len_del and ave_len_dup are in Input/parameters.txt.*

-   Get the median lengths of dels and dups.

    -   Mean is not robust to outliers.

-   Calculate the median / ln(2), which equals to ave_len\_{del, dup} under the assumption of an exponential distribution.

    -   You can compare this value with the direct mean.

### 3) Check the presence of dels and dups in genes you specified as regions of interest.

-   You can get information useful for setting dels and dups in EF.Rint.txt and cloneinit.txt.
